packages/hardhat-core/src/common/bigInt.ts

```diff
@@ -0,0 +1,97 @@
+import type { BigNumber as EthersBigNumberType } from "ethers";
+// eslint-disable-next-line import/no-extraneous-dependencies
+import type { BigNumber as BigNumberJsType } from "bignumber.js";
+// eslint-disable-next-line import/no-extraneous-dependencies
+import type { default as BNType } from "bn.js";
+
+export function normalizeToBigInt(
+  source:
+    | number
+    | bigint
+    | BNType
+    | EthersBigNumberType
+    | BigNumberJsType
+    | string
+): bigint {
+  if (isBigNumber(source)) {
+    return BigInt(source.toString());
+  } else if (
+    typeof source === "string" ||
+    typeof source === "number" ||
+    typeof source === "bigint"
+  ) {
+    if (typeof source === "number") {
+      if (!Number.isInteger(source)) {
+        // eslint-disable-next-line @nomiclabs/hardhat-internal-rules/only-hardhat-error
+        throw new RangeError(
+          `The number ${source} cannot be converted to a BigInt because it is not an integer`
+        );
+      }
+      if (!Number.isSafeInteger(source)) {
+        // eslint-disable-next-line @nomiclabs/hardhat-internal-rules/only-hardhat-error
+        throw new RangeError(
+          `Cannot convert unsafe integer ${source} to BigInt. Consider using BigInt(${source}) instead. For more details, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger`
+        );
+      }
+    }
+    return BigInt(source);
```

Comment:
We should do [exhaustiveness checking](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#exhaustiveness-checking) here.

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/2428/commits/44e5d96974af7755511f929f34008d4fd7a4978c